packages/fxa-auth-server/lib/routes/linked-accounts.ts

```diff
@@ -514,7 +514,21 @@ export class LinkedAccountHandler {
     if (!this.googleAuthClient) {
       throw error.thirdPartyAccountError();
     }
-    const uid = request.auth.credentials.uid;
+    const sessionToken = request.auth && request.auth.credentials;
+    const uid = sessionToken.uid;
+
+    const hasTotpToken = await this.otpUtils.hasTotpToken({ uid });
+
+    // Ensure that the session has the correct assurance level before unlinking
+    if (
+      sessionToken &&
+      hasTotpToken &&
+      (sessionToken.tokenVerificationId ||
+        (sessionToken.authenticatorAssuranceLevel as number) <= 1)
+    ) {
+      throw error.unverifiedSession();
+    }
```

Comment:
Possible to abstract (some of) this AAL check?  Or are the various checks just different enough to make it a pain?

Reply:
I took a stab at doing that refactor and it ended up breaking alot of things :/